src/servers/udp/server.rs

```diff
@@ -235,6 +234,84 @@ impl Drop for ActiveRequests {
     }
 }
 
+/// Wrapper for Tokio [`UdpSocket`][`tokio::net::UdpSocket`] that is bound to a particular socket.
+struct Socket {
+    socket: Arc<tokio::net::UdpSocket>,
+}
+
+impl Socket {
+    async fn new(addr: SocketAddr) -> Result<Self, Box<std::io::Error>> {
+        let socket = tokio::net::UdpSocket::bind(addr).await;
+
+        let socket = match socket {
+            Ok(socket) => socket,
+            Err(e) => Err(e)?,
+        };
+
+        let local_addr = format!("udp://{addr}");
+        tracing::debug!(target: "UDP TRACKER: UdpSocket::new", local_addr, "(bound)");
+
+        Ok(Self {
+            socket: Arc::new(socket),
+        })
+    }
+}
+
+impl Deref for Socket {
+    type Target = tokio::net::UdpSocket;
+
+    fn deref(&self) -> &Self::Target {
+        &self.socket
+    }
+}
+
+impl Debug for Socket {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let local_addr = match self.socket.local_addr() {
+            Ok(socket) => format!("Receiving From: {socket}"),
+            Err(err) => format!("Socket Broken: {err}"),
+        };
+
+        f.debug_struct("UdpSocket").field("addr", &local_addr).finish_non_exhaustive()
+    }
+}
+
+struct Receiver {
+    socket: Arc<UdpSocket>,
+    tracker: Arc<Tracker>,
+    data: RefCell<[u8; MAX_PACKET_SIZE]>,
+}
+
+impl Stream for Receiver {
+    type Item = std::io::Result<AbortHandle>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut buf = *self.data.borrow_mut();
+        let mut buf = tokio::io::ReadBuf::new(&mut buf);
+
+        let Poll::Ready(ready) = self.socket.poll_recv_from(cx, &mut buf) else {
+            return Poll::Pending;
+        };
+
+        let res = match ready {
+            Ok(from) => {
+                let payload = buf.filled().to_vec();
+                let request = UdpRequest { payload, from };
+
+                Some(Ok(tokio::task::spawn(Udp::process_request(
+                    request,
+                    self.tracker.clone(),
+                    self.socket.clone(),
+                ))
+                .abort_handle()))
```

Comment:
@da2ce7 I like this trait implementation. However, I like the idea of having a more "pure" receiver that only returns the request. I would move this spawn to the main loop or maybe the Reciever could have another wrapper `StreamProcessor` or something like that that takes the requests and launches the worker. I think in the future, we could have a limited number of workers instead of spawning a new task, although I guess Tokio handles that.